ShelterModule/Controllers/AnnouncementController.cs

```diff
@@ -0,0 +1,122 @@
+        IPetCommand petCommand, IShelterQuery shelterQuery)
+    {
+        _query = query;
+        _command = command;
+        _petQuery = petQuery;
+        _petCommand = petCommand;
+        _shelterQuery = shelterQuery;
+    }
+
+    /// <summary>
+    ///     Returns an announcement with a given ID
+    /// </summary>
+    /// <param name="id"> ID of the announcement that should be returned </param>
+    /// <returns> Announcement with a given ID </returns>
+    [HttpGet]
+    [Route("{id:guid}")]
+    [ProducesResponseType(typeof(AnnouncementResponse), StatusCodes.Status200OK)]
+    [ProducesResponseType(typeof(ValidationProblemDetails), StatusCodes.Status400BadRequest)]
+    [ProducesResponseType(typeof(ValidationProblemDetails), StatusCodes.Status404NotFound)]
+    public async Task<ActionResult<AnnouncementResponse>> Get(Guid id)
+    {
+        var announcement = await _query.GetByIdAsync(id, HttpContext.RequestAborted);
+        if (announcement is null)
+            return NotFound(new NotFoundResponse
+            {
+                ResourceName = nameof(Announcement),
+                Id = id.ToString()
+            });
+
+        return announcement.ToResponse();
+    }
+
+    /// <summary>
+    ///     Returns announcements based on filters
+    /// </summary>
+    /// <param name="query"> query with parameters to filter announcements </param>
+    /// <returns> List of all announcements matching the filters </returns>
+    [HttpGet]
+    [ProducesResponseType(typeof(IReadOnlyList<AnnouncementResponse>), StatusCodes.Status200OK)]
+    public async Task<IReadOnlyList<AnnouncementResponse>> GetAllFiltered(
+        [FromQuery] GetAllAnnouncementsFilteredQuery query)
+    {
+        return (await _query.GetAllFilteredAsync(query, HttpContext.RequestAborted)).Select(s => s.ToResponse()).
+                                                                                     ToList();
+    }
+
+    /// <summary>
+    ///     Creates new announcement
+    /// </summary>
+    /// <param name="request"> Request received </param>
+    /// <returns> Newly created announcement </returns>
+    [HttpPost]
+    [ProducesResponseType(typeof(AnnouncementResponse), StatusCodes.Status200OK)]
+    [ProducesResponseType(typeof(ValidationProblemDetails), StatusCodes.Status400BadRequest)]
+    public async Task<ActionResult<AnnouncementResponse>> Post(AnnouncementCreationRequest request)
+    {
+        var shelter = await _shelterQuery.GetByIdAsync(request.ShelterId, HttpContext.RequestAborted);
+        if (shelter is null)
+            return BadRequest();
+
+        if (request.PetRequest is null && request.PetId is null)
+            return BadRequest();
+
+        var pet = request.PetId is null
+            ? Pet.FromRequest(request.PetRequest!)
+            : await _petQuery.GetByIdAsync((Guid)request.PetId, HttpContext.RequestAborted);
```

Comment:
Instead of casting to `Guid` (which we can do with any type, with various results), we should get a non-nullable value from the `Value` property. Or we can do something like this:
```
if(request.PetId is {} id)
{
   // id is of type 'Guid' here
}
```

Reply:
Also instead of using `!` to tell the compiler that `request.PetRequest` is not `null`, we could do `?? throw new SomeException(message that tells us what's wrong)` instead. `!` should generally work since we have validation, but if for some reason it does not, we won't know what exatly is failing - we will just have a random `NullReferenceException` in an unexpected place.

Reply:
For an exception I would use either `ArgumentException` or `UnreachableException`, but honestly that's not too important, unless you don't use just `Exception` or some random type with a description that clearly suggests that it should not be thrown here.